### PR TITLE
[IMP] website_slides(_survey): allow to keep user progress after leav…

### PR DESCRIPTION
**Pull request by pko-odoo**
…ing course

PURPOSE

Before this commit when a user left a course, his progress was lost and he had
to retake the whole course from the start. It also allowed people from gaining
karma easily by leaving the course and taking the quiz again and again.
After this commit the user will keep his progress event after leaving the
course and rejoining it later.

SPECIFICATION

When leaving the course the corresponding 'slide.channel.partner' record is
deleted but the 'slide.slide.partner' records are kept unchanged. So when
joining the course, a new 'slide.channel.partner' is created again.
On the other hand, after failing a survey all the records are deleted so the
user has to retake the whole certification course.

LINKS

Task ID : 2199207

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/models/slide_channel.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ class ChannelUsersRelation(models.Model):
     partner_id = fields.Many2one('res.partner', index=True, required=True, ondelete='cascade')
     partner_email = fields.Char(related='partner_id.email', readonly=True)
 
-    def _recompute_completion(self):
+    def _recompute_completion(self, compute_karma_channel_finish=True):
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:26]: Would be good to write a docstring for this methods and add explanation about compute_karma_channel_finish.

File: addons/website_slides/models/slide_channel.py [hunk 2]
```diff
@@ -53,11 +54,17 @@ def _recompute_completion(self):
...
-        Remove attendee from a channel, then also remove slide.slide.partner related to.
-        """
+    @api.model
+    def create(self, values):
+        res = super(ChannelUsersRelation, self).create(values)
+        slide_partners = self.env['slide.slide.partner'].sudo().search([
+            ('channel_id', '=', res.channel_id.id),
+            ('partner_id', '=', res.partner_id.id)
+        ])
+        res.filtered(lambda cp: cp.partner_id in slide_partners.partner_id)._recompute_completion(False)
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:64]: small comment on False ? (just one line to tell : we are doing "that")
Also, name the parameter (as it's a named one)

File: addons/website_slides/models/slide_channel.py [hunk 3]
```diff
@@ -501,12 +508,14 @@ def _add_groups_members(self):
         for channel in self:
             channel._action_add_members(channel.mapped('enroll_group_ids.users.partner_id'))
 
-    def _remove_membership(self, partner_ids):
+    def _remove_membership(self, partner_ids, unlink_slide_partners=False):
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:511]: add explanation in docstring about the new param (for which kind of cases it's used)

File: addons/website_slides/models/slide_channel.py [hunk 4]
```diff
@@ -24,7 +23,14 @@ class ChannelUsersRelation(models.Model):
...
 
-    def _recompute_completion(self):
+    def _recompute_completion(self, compute_karma_channel_finish=True):
+        """
+        Method to recompute channel completion called when the user has finished a slide or
+        when a user join a course after leaving it. In the latter case a parameter is passed
+        to prevent the method to add the amount of karma gained when finishing the course.
+        :param compute_karma_channel_finish:
+        :return:
+        """
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:33]:
```suggestion
        """
        Method to recompute channel completion called when the user has finished a slide or
        when a user join a course after leaving it.
        :param compute_karma_channel_finish: add the amount of karma gained when finishing the course. Is False in the latter case - avoid re-adding already won karma. 
        """
```

File: addons/website_slides/models/slide_channel.py [hunk 5]
```diff
@@ -510,55 +524,19 @@ def _add_groups_members(self):
...
-        ])
-        for partner_channel in channel_completed:
-            channel = partner_channel.channel_id
-            total_karma[partner_channel.partner_id.id] += channel.karma_gen_channel_finish
-
-        return total_karma
-
-    def _remove_membership(self, partner_ids):
-        """ Unlink (!!!) the relationships between the passed partner_ids
+        Unlink (!!!) the relationships between the passed partner_ids
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:529]: unnecessary change

File: addons/website_slides/models/slide_channel.py [hunk 6]
```diff
@@ -510,55 +524,19 @@ def _add_groups_members(self):
...
         and the channels and their slides (done in the unlink of slide.channel.partner model).
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:530]: `and their slides` : That is not correct anymore. Well at least, it's now conditioned by unlink_slide_partners.

File: addons/website_slides/models/slide_channel.py [hunk 7]
```diff
@@ -510,55 +524,19 @@ def _add_groups_members(self):
...
-        Remove earned karma when completed quizz """
+        :param partner_ids:
+        :param unlink_slide_partners: when true unlink all the related slide.slide.partner (used in
+                                    website_slides_survey)
+        :return:
```
Comment (dbeguin) [inline addons/website_slides/models/slide_channel.py:534]: you can remove that line has it returns nothing
If you use :param x:, add docstring for each.
Something like
```
:param partner_ids: partners that leave the course (or that failed the certification)
:param unlink_slide_partners: when true unlink all the related slide.slide.partner (used in website_slides_survey)
```

**Comment by tde-banana-odoo** [comment]
Was overseeded by "new attendee status" recently merged (16.3) .
